### PR TITLE
add file download progress bar, manifest generation & d/l support

### PR DESCRIPTION
**Pull request by ixcat**
draft until: https://github.com/datajoint/djarchive-client/pull/2

### CODE REVIEW

**Review by dimitri-yatsenko — APPROVED** [review]

File: djarchive_client/__init__.py [hunk 1]
```diff
@@ -62,15 +103,246 @@ def client(cls, admin=False):
                        for k in ('endpoint', 'access_key', 'secret_key',
                                  'bucket')}
 
+        if admin and not all(('access_key' in create_args,
+                              'secret_key' in create_args)):
```
Comment (dimitri-yatsenko) [inline djarchive_client/__init__.py:107]:
```suggestion
        if admin and ('access_key' not in create_args or 'secret_key' not in create_args):
```

File: djarchive_client/__init__.py [hunk 2]
```diff
@@ -62,15 +103,246 @@ def client(cls, admin=False):
...
+                if subp not in mani_dat:
+                    msg = 'subpath {} not in manifest'.format(subp)
+                    log.error(msg)
+                    raise FileNotFoundError(msg)
+
+                fp_sz, fp_sha = self._manifest(fp)
+
+                ref_sz, ref_sha = mani_dat[subp]['size'], mani_dat[subp]['sha']
+
+                if not all((fp_sz == ref_sz, fp_sha == ref_sha)):
```
Comment (dimitri-yatsenko) [inline djarchive_client/__init__.py:288]:
```suggestion
                if fp_sz != ref_sz or fp_sha != ref_sha:
```

File: djarchive_client/__init__.py [hunk 3]
```diff
@@ -110,13 +382,14 @@ def _revisions(dataset):
...
 
-        nfound = -1
+        nfound = 0
 
         datasets = (dataset,) if dataset else self.datasets()
 
         for ds in datasets:
-            nfound += 1
-            yield from _revisions(ds)
+            for i, r in enumerate(_revisions(ds), start=1):
```
Comment (dimitri-yatsenko) [inline djarchive_client/__init__.py:390]: You can use the `else` clause of the `for` statement and remove `nfound`.

File: djarchive_client/__init__.py [hunk 4]
```diff
@@ -126,77 +399,181 @@ def _revisions(dataset):
...
+                self.fget_object(spath, lpath, display_progress)
+
+                # check file integrity
+                log.debug('verifying integrity of {}'.format(lpath))
+
+                lsz, lsha = self._manifest(lpath)
+
+                if not all((lsz == mani[ssubp]['size'],
+                            lsha == mani[ssubp]['sha'])):
+
```
Comment (dimitri-yatsenko) [inline djarchive_client/__init__.py:526]:
```suggestion
                if lsz != mani[ssubp]['size'] or lsha != mani[ssubp]['sha']:

```

File: djarchive_client/__init__.py [hunk 5]
```diff
@@ -126,77 +399,181 @@ def _revisions(dataset):
...
+                    oerr += 1
+
+                    log.warning('integrity issue fetching {}'.format(spath))
+
+                # and mark as complete
+                nerr += 1 if oerr else 0
```
Comment (dimitri-yatsenko) [inline djarchive_client/__init__.py:532]:
```suggestion
                nerr += bool(oerr)
```

File: djarchive_client/__init__.py (hunk 4 above)
Comment (ixcat) [reply djarchive_client/__init__.py]: adjusted in c74da65

File: djarchive_client/__init__.py (hunk 2 above)
Comment (ixcat) [reply djarchive_client/__init__.py]: adjusted in c74da65

File: djarchive_client/__init__.py (hunk 5 above)
Comment (ixcat) [reply djarchive_client/__init__.py:534]: leaving as is for now; find original more clear/explicit

**Pull request merged**